cmd/routed-eni-cni-plugin/driver/driver.go

```diff
@@ -267,354 +271,358 @@ func WaitForAddressesToBeStable(ifName string, timeout time.Duration) error {
+		fromContainerRule.Table = rtTable
+		if err := n.netLink.RuleAdd(fromContainerRule); err != nil && !isRuleExistsError(err) {
+			return errors.Wrapf(err, "failed to setup fromContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), rtTable)
+		}
+		log.Debugf("Successfully setup fromContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), rtTable)
 	}
+
 	return nil
 }
 
-// buildRoutesForVlan builds routes required for the vlan link.
-func buildRoutesForVlan(vlanTableID int, vlanIndex int, gw net.IP) []netlink.Route {
-	return []netlink.Route{
-		// Add a direct link route for the pod vlan link only.
-		{
-			LinkIndex: vlanIndex,
-			Dst:       &net.IPNet{IP: gw, Mask: net.CIDRMask(32, 32)},
-			Scope:     netlink.SCOPE_LINK,
-			Table:     vlanTableID,
-		},
-		{
-			LinkIndex: vlanIndex,
-			Dst:       &net.IPNet{IP: net.IPv4zero, Mask: net.CIDRMask(0, 32)},
-			Scope:     netlink.SCOPE_UNIVERSE,
-			Gw:        gw,
-			Table:     vlanTableID,
-		},
+func (n *linuxNetwork) teardownIPBasedContainerRouteRules(containerAddr *net.IPNet, rtTable int, log logger.Logger) error {
+	toContainerRule := n.netLink.NewRule()
+	toContainerRule.Dst = containerAddr
+	toContainerRule.Priority = toContainerRulePriority
+	toContainerRule.Table = unix.RT_TABLE_MAIN
+	if err := n.netLink.RuleDel(toContainerRule); err != nil && !containsNoSuchRule(err) {
+		return errors.Wrapf(err, "failed to delete toContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), "main")
 	}
-}
+	log.Debugf("Successfully deleted toContainer rule, containerAddr=%s, rtTable=%v", containerAddr.String(), "main")
 
-// buildVlanLink builds vlan link for the pod.
-func buildVlanLink(vlanID int, parentIfIndex int, eniMAC string) *netlink.Vlan {
-	la := netlink.NewLinkAttrs()
-	la.Name = fmt.Sprintf("vlan.eth.%d", vlanID)
-	la.ParentIndex = parentIfIndex
-	la.HardwareAddr, _ = net.ParseMAC(eniMAC)
-	return &netlink.Vlan{LinkAttrs: la, VlanId: vlanID}
-}
+	if rtTable != unix.RT_TABLE_MAIN {
+		fromContainerRule := netlink.NewRule()
+		fromContainerRule.Src = containerAddr
+		fromContainerRule.Priority = fromContainerRulePriority
+		fromContainerRule.Table = rtTable
 
-func addContainerRule(netLink netlinkwrapper.NetLink, isToContainer bool, addr *net.IPNet, table int) error {
-	if addr == nil {
-		return errors.New("can't add container rules without an IP address")
-	}
-	containerRule := netLink.NewRule()
-	if isToContainer {
-		// Example: 512:	from all to 10.200.202.222 lookup main
-		containerRule.Dst = addr
-		containerRule.Priority = toContainerRulePriority
-	} else {
-		// Example: 1536:	from 10.200.202.222 to 10.200.0.0/16 lookup 2
-		containerRule.Src = addr
-		containerRule.Priority = fromContainerRulePriority
+		// note: older version CNI sets up multiple CIDR based from container rule, so we recursively delete them to be backwards-compatible.
+		if err := netLinkRuleDelAll(n.netLink, fromContainerRule); err != nil {
```

Comment:
Not following this comment. Here, we are dealing with `from a.b.c.d dev eth1` rule isn't it? How is CIDR playing a part here?

From what I see here - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/routed-eni-cni-plugin/driver/driver.go#L553 , we're only attempting to delete a rule that matches the Pod IP.

Reply:
so previously on older CNI versions, the rules is like
```
from a.b.c.d dev to CIDR1 eth1
from a.b.c.d dev to CIDR2 eth1
```
That's why the code dels rules by SRC. 
I simplified the code to use the new introduced netLinkRuleDelAll to achieve same thing. which removes the hacky ipamd code dependency while remains backwards-compatible. (can delete multiple rules that was setup by older version of CNI)

Reply:
Understood. Makes sense when we're moving from an older version that did setup rules per VPC CIDR.